File: T4/EjemploFragment/app/src/main/java/com/system/develop/ejemplofragment/adaptadores/AdaptadorRecycler.java
package com.system.develop.ejemplofragment.adaptadores;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.system.develop.ejemplofragment.R;
import com.system.develop.ejemplofragment.utils.Persona;

import java.util.ArrayList;

public class AdaptadorRecycler extends RecyclerView.Adapter<AdaptadorRecycler.MiHolder> {

    Context c;
    ArrayList<Persona> lista;
    OnElementoAdaptadorListener listener;

    public AdaptadorRecycler(Context c) {
        this.c = c;
        lista = new ArrayList<>();
        lista.add(new Persona("Nombre1", "Apellido", 224));
        lista.add(new Persona("Nombre2", "Apellido", 224));
        lista.add(new Persona("Nombre3", "Apellido", 224));
        lista.add(new Persona("Nombre4", "Apellido", 224));
        lista.add(new Persona("Nombre5", "Apellido", 224));
        lista.add(new Persona("Nombre6", "Apellido", 224));
        listener = (OnElementoAdaptadorListener) c;
    }


    @NonNull
    @Override
    public MiHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View v = LayoutInflater.from(c).inflate(R.layout.item_recycler, viewGroup, false);
        return new MiHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull MiHolder miHolder, int i) {
        final Persona p = lista.get(i);
        miHolder.getNombre().setText(p.getNombre());
        miHolder.getApellido().setText(p.getApellido());
        miHolder.getNombre().setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                listener.onElementoSelected(p.getTelefono());
            }
        });
    }

    @Override
    public int getItemCount() {
        return lista.size();
    }

    public interface OnElementoAdaptadorListener{
        public void onElementoSelected(int telefono);
    }

    class MiHolder extends RecyclerView.ViewHolder {

        TextView nombre, apellido;

        public MiHolder(@NonNull View itemView) {
            super(itemView);
            nombre = itemView.findViewById(R.id.itemNombre);
            apellido = itemView.findViewById(R.id.itemApellido);
        }

        public TextView getNombre() {
            return nombre;
        }

        public TextView getApellido() {
            return apellido;
        }
    }
}

File: Kotlin/T4/codigos2223/T4_ListasHolder/app/src/main/java/com/develop/t4_listasholder/adapters/AdaptadorRecycler.kt
package com.develop.t4_listasholder.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.TextureView
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.develop.t4_listasholder.R
import com.develop.t4_listasholder.model.Usuario
import com.google.android.material.snackbar.Snackbar

class AdaptadorRecycler(var context: Context, var lista: ArrayList<Usuario>) :
    RecyclerView.Adapter<AdaptadorRecycler.MyHolder>() {

    // 2. Creo un objeto de la interfaz para poder utilizarlo
    private lateinit var listener: OnRecyclerUsuarioListener

    init {
        // OnRecyclerUsuarioListener =  OnRecyclerUsuarioListener
        listener = context as OnRecyclerUsuarioListener;
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyHolder {
        var view: View = LayoutInflater.from(context).inflate(R.layout.item_recycler, parent, false)
        var holder = MyHolder(view);
        return holder
    }

    override fun onBindViewHolder(holder: MyHolder, position: Int) {

        // personaliza cada una de las filas -> holder
        var usuarioFila = lista.get(position);
        holder.nombre.setText(usuarioFila.nombre)
        holder.apellido.setText(usuarioFila.apellido)
        holder.correo.setText(usuarioFila.correo)
        holder.nombre.setOnClickListener{
            //Snackbar.make(holder.nombre,"Pulsado nombre ${usuarioFila.nombre}",Snackbar.LENGTH_SHORT).show()
            // 3. Utilizo el método de la interfaz
            listener.comunicaUsuarioSelected(usuarioFila)
        }

        //holder.nombre.setOnLongClickListener(this)
        holder.nombre.setOnLongClickListener{
            // 3. Utilizo el método de la interfaz
            listener.comunicaUsuarioSelected(usuarioFila,position)
            return@setOnLongClickListener true
        }
        holder.apellido.setOnClickListener{
            Snackbar.make(holder.nombre,"Pulsado apellido",Snackbar.LENGTH_SHORT).show()
        }
        holder.correo.setOnClickListener{
            Snackbar.make(holder.nombre,"Pulsado correo",Snackbar.LENGTH_SHORT).show()
        }

        // boton.setOnClick()-->cambiar de pantalla
        /*usuarioFila.apellido
        usuarioFila.nombre
        usuarioFila.correo*/

    }

    override fun getItemCount(): Int {
        return lista.size
    }

    // 1. Origen de los datos creo una interfaz

    interface OnRecyclerUsuarioListener{
        fun comunicaUsuarioSelected(usuario: Usuario)
        fun comunicaUsuarioSelected(usuario: Usuario, posicion: Int)
    }


    // clase anidada
    inner class MyHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var nombre: TextView;
        var correo: TextView;
        var apellido: TextView;

        init {
            nombre = itemView.findViewById(R.id.nombre_fila)
            correo = itemView.findViewById(R.id.correo_fila)
            apellido = itemView.findViewById(R.id.apellido_fila)
        }


    }



}